### ISSUE TITLE
Serial library wont compile on Windows

### ISSUE THREAD

**Issue by accaldwell** [euantorano/serial.nim#24]
Attempting to compile any program that imports serial is failing on Windows:

`C:\Users\Adam\.nimble\pkgs\serial-#head\serial\private\ffi\ffi_windows.nim(29, 27) Error: cannot convert 4294967295 to DWORD`

I think maybe the specified code should be changed to:

`MAXDWORD*: DWORD = DWORD(high(int32))`

since winlean defines a DWORD as being of type int32. But, I might be misunderstanding something.

Either way, that change gets rid of the above error, but leads to further errors about invalid else clauses. Commenting out those just gets to yet another error:

`C:\Users\Adam\.nimble\pkgs\serial-#head\serial\private\utils\list_serialports_windows.nim(65, 26) Error: -1 can't be converted to HKEY`

HKEY is a uint, so the error makes sense, but I'm unsure what exactly you were doing originally here, so that's as far as I got.

I'm using nim 1.0.0 and serial 1da4cc14b19e0248a966c6d808093af31414bf2b.

**Comment by euantorano** [comment]
Thanks for the report. I'll see what I can do this week to fix this. I did originally test on Windows, but it's been a while.

**Comment by euantorano** [comment]
Regarding issue #2, `HKEY` is defined in Windows as the same as `HANDLE` which is defined as `PVOID` which is defined as `typedef void *PVOID;`.

It looks like the `HKEY` definition is incorrect, which would cause problems.

**Comment by euantorano** [comment]
Additionally, a `DWORD` should actually be a 32 bit unsigned integer according to microsoft: https://docs.microsoft.com/en-us/windows/win32/winprog/windows-data-types

Nim seems to get this wrong...

**Comment by accaldwell** [comment]
Looks like the DWORD issue is here: https://github.com/nim-lang/Nim/issues/9848

**Comment by euantorano** [comment]
Thanks, I've also created an issue about the `HANDLE` definition and mentioned it on IRC: https://github.com/nim-lang/Nim/issues/12327

I'll create a PR to Nim in a moment and see how much stuff breaks when I change the definitions 🤣

**Comment by LucaWolf** [comment]
Well, in my view MS is not consistent either. All those DWORD and stuff seem derived from pointer to void type. In C, pointer to void has no sign, so is "free for all". Then MS defines variables of those types in the unsigned range but inconsistently also return -1 for some API. So in their view pointer to void could take both signed and unsigned values. Schrödinger would be very pleased :-)

I have circumvented this issue by casting signed <-> unsigned where required as:
```nim
# be careful with MAXDWORD as a wrong value induces huge delays on receiving timers
# the correct value should be the binary representation of max unsigned int 
# 0xFFFFFFFF (at least on 32 bit platform, not very sure about 64bit)
MAXDWORD*: DWORD = cast[DWORD](high(uint32)) # in ffi_windows.nim
if regKey != cast[HKEY](-1): # in list_serialports_windows.nim
```

Also, Nim 1.0.0 warns about unused modules (stream, os and events in various nim files).

Most importantly, if you want to cross compile to MS (i.e. develop on Linux host but target Windows builds with mingw) please lower case the "Windows.h" header of the FFI declarations. This would make the module compile on both Windows (case insensitive) and Linux (MinGw defined the header lower case and linux itself is case sensitive). Ex:
``` nim
#CommProp* {.importc: "COMMPROP", header: "Windows.h", incompleteStruct.} = object
CommProp* {.importc: "COMMPROP", header: "windows.h", incompleteStruct.} = object
```
The above worked fine for my purpose.

**Comment by mp035** [comment]
I had all of the above issues and a few others to get compilation working with Nim 1.0.2.  As a temporary workaround, users are welcome to use my fork until nim fixes its bugs in the windows module.  I have just updated it to use @LucaWolf 'cast' solutions instead of changing types which was my original solution.  
@euantorano do you want me to issue a pull request, or would you prefer to wait until Nim fixes its windows types?  The changes using cast should still work once Nim fixes are pused.  Some of my other changes (removal of invalid else cases) are as required a result of the 1.0 release, I beleive.

**Comment by euantorano** [comment]
A PR would be great. I started looking at changing the Nim types and it was going to require a lot of changes all over the standard library so I doubt it’ll be changed any time soon.

**Comment by euantorano** [comment]
Fixed in #29
